server/src/main/java/org/elasticsearch/index/search/QueryStringQueryParser.java

```diff
@@ -97,6 +100,9 @@
     private MultiTermQuery.RewriteMethod fuzzyRewriteMethod;
     private boolean fuzzyTranspositions = FuzzyQuery.defaultTranspositions;
 
+    // set of field names this parser targets, used to check limits on expanded field names
+    private Set<String> queriedFields = new HashSet<>();
```

Comment:
I don't think this is needed. We can check the limit in `QueryStringQueryParser#extractMultiFields`, which is called every time we parse a block.

Reply:
My understanding might be off here, but my assumption was that "extractMultiFields" gets called for every block, so e.g. for a query string like "f1:foo f2:bar f3:baz" it gets called three different times for one field each call. In order to get the number of fields really queried we'd need to at least have a counter somewhere, and in oder to check for duplicates (e.g. "f1:foo f1:bar f2:baz" which should only count for two fields) we'd need something to store the field names already used. This is why I introduced this string set here thats filled with each call to `extractMultiFields`. Does this make sense or what am I missing?

Reply:
I see what you mean but I am not sure that we need to sum all fields. We want to check the limit on each block eagerly, not the total of clauses.

Reply:
Ok, just to re-check my understanding, because I might have thought about a different goal in this fix. Assuming we have `indices.query.bool.max_clause_count` set to 50 and query string like "foo*:somevalue bar*:otherValue", and both expressions expand to 40 mapped fields, would we want to accept or reject that? In the end we are targetting 80 fields, so my assumption was that we'd want to sum here, but I'm also fine with another interpretation. Although that can easily lead to the number of fields that are hit by a search can be much larger than the limit?

Reply:
The goal of this change is to only check the limit per block. So in your case, if the limit is 50 then the parsing would not throw any error since all blocks are within the limit.

Reply:
Thanks for the clarification, this is indeed different from what I was aiming at. I will adapt the approach and the testing consequently.